File: ports/jagentic-core/src/test/java/org/jagentic/core/FeaturesTest.java
package org.jagentic.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

import org.jagentic.core.llm.ChatResult;
import org.jagentic.core.llm.LlmBrain;
import org.jagentic.core.llm.StubChatClient;
import org.jagentic.core.skill.Skill;
import org.jagentic.core.skill.SkillRegistry;
import org.jagentic.core.structured.Structured;

/** Phase B feature tests: skills, structured output, richer listeners. */
class FeaturesTest {

  private AgentContext ctx(ToolRegistry tools) {
    return new AgentContext("c1", "alice", new ConversationStore.InMemory(),
        new KeyedStateStore.InMemory(), tools, null);
  }

  // ---- skills ----

  @Test
  void skillRegistryExpands() {
    SkillRegistry reg = new SkillRegistry().register(
        new Skill("billing", List.of("get_balance", "refund"), "Be precise about amounts.", List.of("account")));
    SkillRegistry.Expanded e = reg.expand(List.of("billing"));
    assertEquals(List.of("get_balance", "refund"), e.tools());
    assertTrue(e.promptFragment().contains("precise"));
    assertEquals(List.of("account"), e.facts());
  }

  // ---- structured output ----

  @Test
  void validateRequiredAndTypes() {
    Map<String, Object> schema = Map.of("type", "object", "required", List.of("category", "amount"),
        "properties", Map.of("category", Map.of("type", "string"), "amount", Map.of("type", "number")));
    assertTrue(Structured.validate(Map.of("category", "refund", "amount", 42.0), schema).isEmpty());
    assertTrue(Structured.validate(Map.of("category", "refund"), schema).stream().anyMatch(s -> s.contains("amount")));
    assertEquals(2, Structured.validate(Map.of("category", 1, "amount", "x"), schema).size());
  }

  @Test
  void parseStructuredToleratesProse() {
    Map<String, Object> schema = Map.of("type", "object", "required", List.of("ok"),
        "properties", Map.of("ok", Map.of("type", "boolean")));
    Structured.Result r = Structured.parse("here you go: {\"ok\": true} done", schema);
    assertTrue(r.ok());
    assertEquals(Boolean.TRUE, r.value().get("ok"));
  }

  @Test
  void llmBrainOutputSchemaReturnsValidatedJson() {
    Map<String, Object> schema = Map.of("type", "object", "required", List.of("category"),
        "properties", Map.of("category", Map.of("type", "string")));
    LlmBrain brain = new LlmBrain(new StubChatClient(List.of(ChatResult.text("{\"category\": \"refund\"}"))), "triage")
        .withOutputSchema(schema);
    TurnResult res = new Agent("triage", "p", brain).turn(new Event("c1", "u", "refund please"), ctx(new ToolRegistry()));
    assertEquals("[triage] {\"category\":\"refund\"}", res.reply);
  }

  // ---- richer listeners ----

  @Test
  void toolCallAndCompositeListenerHooksFire() {
    MetricsListener m1 = new MetricsListener();
    MetricsListener m2 = new MetricsListener();
    CompositeListener composite = new CompositeListener(m1, m2);
    ToolRegistry tools = new ToolRegistry().register("get_balance", "balance", p -> 1234.56);
    Agent pay = new Agent("payments", "p", new LlmBrain(new StubChatClient(List.of(
        ChatResult.toolCall("get_balance", Map.of()), ChatResult.text("done"))), "payments"));
    RoutedGraph graph = new RoutedGraph((ev, c) -> "payments", Map.of("payments", pay), null,
        List.of(), List.of(composite));
    graph.handle(new Event("c1", "u", "balance?"), ctx(tools));
    for (MetricsListener m : List.of(m1, m2)) {
      assertEquals(1, m.turns.get());
      assertEquals(1, m.toolCalls.get());
      assertEquals(1, m.pathCount("payments"));
    }
  }

  @Test
  void listenerErrorHookOnToolFailure() {
    MetricsListener m = new MetricsListener();
    ToolRegistry tools = new ToolRegistry().register("boom", "fails", p -> { throw new RuntimeException("boom"); });
    Agent pay = new Agent("payments", "p", new LlmBrain(new StubChatClient(List.of(
        ChatResult.toolCall("boom", Map.of()), ChatResult.text("x"))), "payments"));
    RoutedGraph graph = new RoutedGraph((ev, c) -> "payments", Map.of("payments", pay), null,
        List.of(), List.of(m));
    try {
      graph.handle(new Event("c1", "u", "go"), ctx(tools));
    } catch (RuntimeException ignore) {
    }
    assertEquals(1, m.errors.get());
  }
}

File: ports/experimental/go/core/features_test.go
package core

import (
	"strings"
	"testing"
)

// ---- skills ----

func TestSkillRegistryExpands(t *testing.T) {
	reg := NewSkillRegistry().Register(Skill{
		Name: "billing", Tools: []string{"get_balance", "refund"},
		PromptFragment: "Be precise about amounts.", RequiredFacts: []string{"account"}})
	tools, fragment, facts := reg.Expand([]string{"billing"})
	if len(tools) != 2 || tools[0] != "get_balance" {
		t.Fatalf("tools = %v", tools)
	}
	if !strings.Contains(fragment, "precise") || len(facts) != 1 {
		t.Fatalf("fragment/facts wrong: %q %v", fragment, facts)
	}
}

// ---- structured output ----

func TestValidateSchema(t *testing.T) {
	schema := map[string]any{"type": "object", "required": []any{"category", "amount"},
		"properties": map[string]any{"category": map[string]any{"type": "string"},
			"amount": map[string]any{"type": "number"}}}
	if errs := ValidateSchema(map[string]any{"category": "refund", "amount": 42.0}, schema); len(errs) != 0 {
		t.Fatalf("valid object rejected: %v", errs)
	}
	if errs := ValidateSchema(map[string]any{"category": "refund"}, schema); len(errs) == 0 {
		t.Fatal("missing field not caught")
	}
	if errs := ValidateSchema(map[string]any{"category": 1.0, "amount": "x"}, schema); len(errs) != 2 {
		t.Fatalf("type errors = %v", errs)
	}
}

func TestParseStructuredToleratesProse(t *testing.T) {
	schema := map[string]any{"type": "object", "required": []any{"ok"},
		"properties": map[string]any{"ok": map[string]any{"type": "boolean"}}}
	obj, errs := ParseStructured(`here you go: {"ok": true} done`, schema)
	if len(errs) != 0 || obj["ok"] != true {
		t.Fatalf("parse failed: %v %v", obj, errs)
	}
}

func TestLlmBrainOutputSchemaReturnsValidatedJSON(t *testing.T) {
	schema := map[string]any{"type": "object", "required": []any{"category"},
		"properties": map[string]any{"category": map[string]any{"type": "string"}}}
	brain := NewLlmBrain(NewStubChatClient(TextResult(`{"category":"refund"}`)), "triage", "", nil, 6).
		WithOutputSchema(schema)
	res := NewAgent("triage", "p", brain).Turn(NewEvent("c1", "u", "refund please"), llmCtx(NewToolRegistry()))
	if res.Reply != `[triage] {"category":"refund"}` {
		t.Fatalf("reply = %q", res.Reply)
	}
}

// ---- richer listeners ----

func TestToolCallAndCompositeListenerHooksFire(t *testing.T) {
	m1, m2 := NewMetricsListener(), NewMetricsListener()
	composite := NewCompositeListener(m1, m2)
	tools := NewToolRegistry().Register("get_balance", "balance", func(p map[string]any) any { return 1234.56 })
	pay := NewAgent("payments", "p", NewLlmBrain(NewStubChatClient(
		ToolCall("get_balance", nil), TextResult("done")), "payments", "", nil, 6))
	graph := NewRoutedGraph(func(ev Event, ctx *AgentContext) string { return "payments" },
		map[string]*Agent{"payments": pay}, nil).WithListeners(composite)
	graph.Handle(NewEvent("c1", "u", "balance?"), llmCtx(tools))
	for _, m := range []*MetricsListener{m1, m2} {
		if m.Turns != 1 || m.ToolCalls != 1 || m.PathCount("payments") != 1 {
			t.Fatalf("metrics = turns:%d tools:%d paths:%v", m.Turns, m.ToolCalls, m.Paths)
		}
	}
}

func TestListenerErrorHookOnToolPanic(t *testing.T) {
	m := NewMetricsListener()
	tools := NewToolRegistry().Register("boom", "fails", func(p map[string]any) any { panic("tool exploded") })
	pay := NewAgent("payments", "p", NewLlmBrain(NewStubChatClient(
		ToolCall("boom", nil), TextResult("x")), "payments", "", nil, 6))
	graph := NewRoutedGraph(func(ev Event, ctx *AgentContext) string { return "payments" },
		map[string]*Agent{"payments": pay}, nil).WithListeners(m)
	func() {
		defer func() { recover() }()
		graph.Handle(NewEvent("c1", "u", "go"), llmCtx(tools))
	}()
	if m.Errors != 1 {
		t.Fatalf("errors = %d", m.Errors)
	}
}
